flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -130,95 +130,108 @@ public void checkpointState(
-					checkpointMetrics.getSyncDurationMillis());
-			}
+		if (LOG.isDebugEnabled()) {
+			LOG.debug(
+				"{} - did NOT finish synchronous part of checkpoint {}. Alignment duration: {} ms, snapshot duration {} ms",
+				taskName, metadata.getCheckpointId(),
+				metrics.getAlignmentDurationNanos() / 1_000_000,
+				metrics.getSyncDurationMillis());
+		}
 
-			if (checkpointOptions.getCheckpointType().isSynchronous()) {
-				// in the case of a synchronous checkpoint, we always rethrow the exception,
-				// so that the task fails.
-				// this is because the intention is always to stop the job after this checkpointing
-				// operation, and without the failure, the task would go back to normal execution.
-				throw ex;
-			} else {
-				env.declineCheckpoint(checkpointMetaData.getCheckpointId(), ex);
-			}
+		if (options.getCheckpointType().isSynchronous()) {
+			// in the case of a synchronous checkpoint, we always rethrow the exception,
+			// so that the task fails.
+			// this is because the intention is always to stop the job after this checkpointing
+			// operation, and without the failure, the task would go back to normal execution.
+			throw ex;
+		} else {
+			env.declineCheckpoint(metadata.getCheckpointId(), ex);
+		}
+	}
+
+	private void finishAndReportAsync(Map<OperatorID, OperatorSnapshotFutures> snapshotFutures, CheckpointMetaData metadata, CheckpointMetrics metrics) {
+		// we are transferring ownership over snapshotInProgressList for cleanup to the thread, active on submit
+		executorService.execute(new AsyncCheckpointRunnable(
+			snapshotFutures,
+			metadata,
+			metrics,
+			System.nanoTime(),
+			taskName,
+			closeableRegistry,
+			env,
+			asyncExceptionHandler));
+	}
+
+	private void takeSnapshotSync(
+		Map<OperatorID, OperatorSnapshotFutures> operatorSnapshotsInProgress, CheckpointMetaData checkpointMetaData,
+		CheckpointMetrics checkpointMetrics, CheckpointOptions checkpointOptions,
+		OperatorChain<?, ?> operatorChain,
+		Supplier<Boolean> isCanceled) throws Exception {
+
+		long checkpointId = checkpointMetaData.getCheckpointId();
+		long started = System.nanoTime();
+
+		ChannelStateWriteResult channelStateWriteResult = checkpointOptions.getCheckpointType() == CHECKPOINT ?
+								channelStateWriter.getWriteResult(checkpointId) :
+								ChannelStateWriteResult.EMPTY;
+
+		CheckpointStreamFactory storage = checkpointStorage.resolveCheckpointStorageLocation(checkpointId, checkpointOptions.getTargetLocation());
+
+		for (StreamOperatorWrapper<?, ?> operatorWrapper : operatorChain.getAllOperators(true)) {
+			operatorSnapshotsInProgress.put(
+				operatorWrapper.getStreamOperator().getOperatorID(),
+				buildOperatorSnapshotFutures(
+					checkpointMetaData,
+					checkpointOptions,
+					operatorChain,
+					operatorWrapper.getStreamOperator(),
+					isCanceled,
+					channelStateWriteResult,
+					storage));
 		}
+
+		checkpointStorage.clearCacheFor(checkpointId);
```

Comment:
If `buildOperatorSnapshotFutures` encounters exception, the clear will not be executed, should place within `finally`?

Reply:
Yes, will fix this.